File: aoc/aoc2022/4/day4.py
#!/usr/bin/env python3

import sys
import re

# range() in python is *exclusive* of the stop value
# In this challenge, the input ranges are *inclusive*, and using python's
# builtin range() makes this a little difficult.
# So, write my own inclusive range class
class InclusiveRange:
    def __init__(self, start, stop):
        self.start = start
        self.stop = stop

    def __str__(self):
        return "{}-{}".format(self.start, self.stop)

    """ 
    Check if one range fully contains another range.

    This method allows python's `in` operator to work.
    InclusiveRange and `int` types are supported.
    """
    def __contains__(self, other):
        if isinstance(other, self.__class__):
            # Is the 'other' range fully inside of our range?
            return self.start >= other.start and self.stop <= other.stop
        elif isinstance(other, int):
            # Is a number in the range?
            return self.start <= other and self.stop >= other
        else:
            raise Exception("Unexpected type: {}".format(type(other)))

input_re = re.compile(r'(\d+)-(\d+),(\d+)-(\d+)')

def parse(input):
    input.rstrip()
    m = input_re.match(input)
    if m:
        elf1 = InclusiveRange(int(m.group(1)), int(m.group(2)))
        elf2 = InclusiveRange(int(m.group(3)), int(m.group(4)))
        return elf1, elf2
    else:
        raise Exception(format("Some bug parsing input? Line was: {}", input))

def full_overlap(one, two):
    return one in two or two in one

def any_overlap(one, two):
    if two.start in one or two.stop in one:
        return True
    if one.start in two or one.stop in two:
        return True
    return False

count_full_overlap = 0
count_any_overlap = 0

for line in [x.rstrip() for x in sys.stdin]:
    elf1, elf2 = parse(line)

    if full_overlap(elf1, elf2):
        count_full_overlap += 1
    if any_overlap(elf1, elf2):
        count_any_overlap += 1

print("Full overlap: {}".format(count_full_overlap))
print("Any overlap: {}".format(count_any_overlap))

File: aoc/aoc2022/4/day4.rb
#/usr/bin/env ruby

class Reader
    include Enumerable

    def each
        while !ARGF.eof?
            yield ARGF.readline.chomp
        end
    end
end

full_overlap = 0
any_overlap = 0
Reader.new.each do |line|
    # Parse with the format with a regex.
    # Uses capture groups in order to pick each number out of the line.
    m = line.match(/(\d+)-(\d+),(\d+)-(\d+)/)
    raise "Bug?" unless m

    # Turn each text range "1-3" into a Range object 1..3
    elf1, elf2 = Range.new(*m[1..2].map(&:to_i)), Range.new(*m[3..4].map(&:to_i))

    full_overlap += 1 if elf1.cover?(elf2) || elf2.cover?(elf1)

    if elf1.cover?(elf2.min) || elf1.cover?(elf2.max) \
        || elf2.cover?(elf1.min) || elf2.cover?(elf1.max)
        any_overlap += 1
    end
end

puts "Full overlap: #{full_overlap}"
puts "Any overlap: #{any_overlap}"
